Edit cli readme

jtoar opened the pull request:
This PR improves the cli readme. The cli&mdash;the generators especially&mdash;gets a lot of PR-traffic, so I'm hoping this will be helpful. I'll continue updating this doc when things are in the works/get merged, like #653.

I wanted to go into more detail, but it's already much longer than I anticipated. :sweat_smile: 

Some specific questions (to no one in particular):

### Purpose and Vision

Does it sound too airy here? It also doesn't talk too much about the future, just that we'll keep adding more stuff, and that stuff will probably be side/target specific.

### Package Lead

Who should I list? Right now I list @cannikin and @peterp.

### Roadmap

Would appreciate any additions here!

### FAQ

Like the roadmap, anything else to add here? There's only two questions right now:

```markdown
## FAQ

### I want to alias `yarn rw`

You're not the only one. See the discussion [here](https://github.com/redwoodjs/redwood/issues/328).

### Can I customize the generators?

Not yet, but we're talking about it! See the ongoing dicussions in these issues:

- Investigate integrating or replacing generators with Plop [#653](https://github.com/redwoodjs/redwood/issues/653)
- BYO Components to Scaffold Generator [#473](https://github.com/redwoodjs/redwood/issues/473)
```

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+Redwood provides a first-class CLI that helps you at every stage of development, from your first commit to your first deploy. And it comes with Redwood, which means no extra software to install!
 
-Be sure to prefix all Redwood CLI commands with `yarn`. For example, `yarn redwood new`.
+Redwood uses [yarn workspaces](https://yarnpkg.com/features/workspaces) to separate your app's sides, generators give you a smooth DX, and Prisma to manage your database. We have a generator for nearly everything, on both sides of the app, from components to functions.
```
```suggestion
Redwood uses [yarn workspaces](https://yarnpkg.com/features/workspaces) to separate your app's sides, generators that give you a smooth DX, and Prisma to manage your database. We have a generator for nearly everything, on both sides of the app, from components to functions.
```

jtoar replied:
Thanks I totally missed this. Thinking we'll go with "to" for parallelism.

Tobbe commented on packages/cli/README.md:
````diff
@@ -1,66 +1,478 @@
+  .argv
+```
+
+you'll see the arguments to the `command`  method spread across named, exported constants, like:
````
```suggestion
you'll see the arguments to the `command` method spread across named and exported constants, like:
```

jtoar replied:
Thinking we'll just take out "named" here since that part is starting to seem obvious.

Tobbe commented on packages/cli/README.md:
````diff
@@ -1,66 +1,478 @@
+}
+```
+
+It helps to know this before consulting the yargs doc, as most of the it's written in the former api instead of the latter.
````
```suggestion
It helps to know this before consulting the yargs doc, as most of it is written using the former api instead of the latter.
```

jtoar replied:
+1 for "in -> using". For "it's -> it is", contractions make the writing more conversational, although I'm guilty of misusing them. Motion to keep the contraction?

Tobbe replied:
I agree that the contractions make it feel more conversational and friendly. That's why I kept most of them. But sometimes I feel they make the text more difficult to read. This is one of those places. However, I'm not a native English speaker, so I'll leave this to someone else to settle 👨‍⚖️

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+- **descriptions shouldn't end in periods:** Again, just a stylistic choice&mdash;but stick to it!
+- **`builder` should be a function:** This enables the positional api (more on that later)
+
+If none of these make sense yet, don't worry! You'll see them come up in the next section where we you walk through adding a command.
```
```suggestion
If none of these make sense yet, don't worry! You'll see them come up in the next section where we walk you through adding a command.
```

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+<!-- TODO -->
+`'build'`, the first word in the string, specifies the name of the command, and `'[side..]'` indicates that `build` takes an optional positional argument (named `side`--relevant for the `builder` and the `handler`). The dots (`..`) trailing `side` indicate that you can provide an array of strings (see [Variadic Positional Arguments](https://github.com/yargs/yargs/blob/master/docs/advanced.md#vari ...
+
+<!-- The square brakcets indicate that the `build` takes an optional positional argument named `side` (relevant for `builder` and `handler`). The dots (`..`) trailing `side` indicate that you can provide an array of strings (see [Variadic Positional Arguments](https://github.com/yargs/yargs/blob/master/docs/advanced.md#variadic-positional-arguments)), for example: -->
```
```suggestion
<!-- The square brackets indicate that `build` takes an optional positional argument named `side` (relevant for `builder` and `handler`). The dots (`..`) trailing `side` indicate that you can provide an array of strings (see [Variadic Positional Arguments](https://github.com/yargs/yargs/blob/master/docs/advanced.md#variadic-positional-arguments)), for example: -->
```

Tobbe commented on packages/cli/README.md:
````diff
@@ -1,66 +1,478 @@
+...
+```
+
+Other than that, there's not much more to it. You should of course make the description descriptive. And for consistency, don't punctuate the it. I.e., we prefer this
````
```suggestion
Other than that, there's not much more to it. You should of course make the description descriptive. And for consistency, don't punctuate it. I.e., we prefer this
```

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+
+`builder` configures the positional arguments and options for the command. 
+
+While `builder` can be an object, the [positional argument api](https://yargs.js.org/docs/#api-positionalkey-opt) is only available if buidler's a function, so that's what we go with. We'll discuss more about builder being an object when we talk about testing.
```
```suggestion
While `builder` can be an object, the [positional argument api](https://yargs.js.org/docs/#api-positionalkey-opt) is only available if builder's a function, so that's what we go with. We'll discuss more about builder being an object when we talk about testing.
```

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+
+You should always specify the `description` and `type` properties. The rest depends. But generally, if you can fill out more, you should.
+
+While `side` would've worked in a bare-bones kind of way if we didn't do this, `builder`'s the only way to let the command know about it's options (only showing the relevant bits here):
```
```suggestion
While `side` would've worked in a bare-bones kind of way if we didn't do this, `builder`'s the only way to let the command know about its options (only showing the relevant bits here):
```

Tobbe commented on packages/cli/README.md:
````diff
@@ -1,66 +1,478 @@
+}
+```
+
+The logic that goes in a command's `handler`'s varies too much to comment on generally. But you'll see similarities among the handlers of commands that do similar things, like generators. 
````
```suggestion
The logic that goes in a command's `handler` varies too much to comment on generally. But you'll see similarities among the handlers of commands that do similar things, like generators. 
```

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+
+<!-- ### builder and testing
+
+While we want you to export `builder` as a function, you can use a `builderObject` to define the ags/options.
```
```suggestion
While we want you to export `builder` as a function, you can use a `builderObject` to define the arguments/options.
```

Tobbe commented on packages/cli/README.md:
````diff
@@ -1,66 +1,478 @@
+└── __tests__
+```
+
+There's files and directories here that aren't yargs related&mdash;`README.md`, `helper.js`, and `__tests__`, but because yargs will only use the files that export the appropriate constants, it's ok.
````
```suggestion
There are files and directories here that aren't yargs related&mdash;`README.md`, `helper.js`, and `__tests__`, but because yargs will only use the files that export the appropriate constants, that's ok.
```

Tobbe commented on packages/cli/README.md:
````diff
@@ -1,66 +1,478 @@
 ```
 
-## Development
+Since a typical generator writes files, needs templates to do so, and needs tests to ensure it works, we use this command-in-a-directory structure to keep things organized. The entry-point command, `generate.js`, will stil be able to find these command files (here `page.js`) because of the second argument (`{ recurse: true }`) to `commandDir`.
````
```suggestion
Since a typical generator writes files, needs templates to do so, and needs tests to ensure it works, we use this command-in-a-directory structure to keep things organized. The entry-point command, `generate.js`, will still be able to find these command files (here `page.js`) because of the second argument (`{ recurse: true }`) to `commandDir`.
```

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,478 @@
+If you're adding a new generator or modifying an existing one, you're gonna wanna test it! (Well, at least we want you to.)
+
+<!-- inc -->
+We use [Jest](https://jestjs.io/docs/en/getting-started.html), so you'll want to be familiar with it's api.
```
```suggestion
We use [Jest](https://jestjs.io/docs/en/getting-started.html), so you'll want to be familiar with its api.
```

jtoar commented:
@Tobbe I super appreciate the review! I feel bad because some of the stuff you edited was commented out.

Would you like for me to request a review from you when I think it's done?

Tobbe commented:
> @Tobbe I super appreciate the review! I feel bad because some of the stuff you edited was commented out.
> 
> Would you like for me to request a review from you when I think it's done?

No need to feel bad. That's a lot of documentation you're contributing here, and that's great! When/if I want to tinker with the generators I'm sure I will reference this a lot 🙂 

Sure, I'll read through it again when you feel it's ready 📖

jtoar commented:
@Tobbe If you're up for it, this is ready for review. 

Don't feel like you need to read all of it! It's longer than I anticipated. But one thing I'd definitely like to know is if you find the section on generators helpful? You mentioned you were interested in tinkering with them and would use this as a reference to do so, so I want to make sure it checks that box.

Tobbe commented on packages/cli/README.md:
```diff
@@ -1,66 +1,739 @@
+
+If you're converting a generator, read the _Goals_ section of tracking issue [#523](https://github.com/redwoodjs/redwood/issues/523); it details some specs you should comply with.
+
+Some of the generators have already have been converted; use them as a reference (linking to the PRs here):
```
```suggestion
Some of the generators have already been converted; use them as a reference (linking to the PRs here):
```

jtoar replied:
Nice catch!

Tobbe commented:
@jtoar I'm impressed! A lot of work, and great documentation! :)  Except for my minor comment above I think this is awesome. And yes, the section on the generators is helpful 🙂  +1

jtoar commented:
@Tobbe Awesome! Really appreciate the time you spent reviewing it, you fixed a ton of typos that I just wouldn't of caught.

thedavidprice commented on packages/cli/README.md:
```diff
@@ -1,66 +1,739 @@
+Since the CLI is the entry point to Redwood, as Redwood continues to grow&mdash;especially as we add more sides and targets&mdash;so will the CLI.
 
-## Command line basics
+## Package Lead
```
```suggestion
## Package Leads
```

thedavidprice commented on packages/cli/README.md:
```diff
@@ -1,66 +1,739 @@
+
+- [@cannikin](https://github.com/cannikin)
+- [@peterp](https://github.com/peterp)
+
```
```suggestion
[@thedavidprice](https://github.com/thedavidprice)
```

thedavidprice commented:
@jtoar This is just fantastic. I ❤️ the details, as will anyone and everyone who works on CLI in the future.

Misc items:
- this replaces the [generator specific README](https://github.com/redwoodjs/redwood/blob/master/packages/cli/src/commands/generate/README.md), correct? If so, go ahead and delete it with this PR.
- anytime a Command is changed or added, there needs to be correlating updates to Tests and Docs (both devolving CLI doc and anything else, e.g. Auth). Probably include this in the Contributing "Overview" or "Best Practices" section.
    - note on tests: test coverage isn't 100% but it's in progress -- https://github.com/redwoodjs/redwood/issues/682
- Tests for Destroy command is different and relies on the `__Mocks__` Hoping to get some better integration with #682, but would be worth adding a section to Contributing specific to Destroy tests.

thedavidprice commented on packages/cli/README.md:
```diff
@@ -44,7 +44,7 @@ Since the CLI is the entry point to Redwood, as Redwood continues to grow&mdash;
 - [@cannikin](https://github.com/cannikin)
 - [@peterp](https://github.com/peterp)
-
+[@thedavidprice](https://github.com/thedavidprice)
```
```suggestion
- [@thedavidprice](https://github.com/thedavidprice)
```

jtoar commented:
@thedavidprice Will take a look at #682 and queue up the other two misc items.

jtoar commented:
@thedavidprice Should be ready to go. Just to recap the changes:
- Added a destroyers section (as a subsection to generators, it includes information about testing destroyers).
- Mentioned updating docs and tests as a part of best practices.
- Merged and deleted the auth generator readme into this doc.

peterp commented on packages/cli/README.md:
```diff
@@ -1,66 +1,789 @@
+To make a command using the advanced api, yargs requires that you export [four constants](https://github.com/yargs/yargs/blob/master/docs/advanced.md#providing-a-command-module):
 
-We create a new Redwood application by running the `yarn create redwood-app <project-dir>` command, where `<project-dir>` is the path of the to-be Redwood project.
+|Constant|Type|Description|
+|:-|:-|:-|
+|`command`|string|A yargs command definition. You specify the command's name and arguments here|
+|`description`|string|A description of the command; shown in the help message|
+|`builder`|function|Effectively "builds" the command's arguments/options|
+|`handler`|function|The function invoked by the command; does all the work|
```
I think there's a markdown formatter available in vscode that'll make these easier to read.

jtoar replied:
Thanks, just did that. Looks so much better. I 'll have to do that across the docs/from now on.

Turns out I had the extension already but wasn't using the keyboard shortcut to trigger the formatter.

thedavidprice commented:
🚀
